### PR TITLE
Add integration tests running on RHEL8 to project pipeline

### LINKED ISSUE #181: Release automation - Run integration tests on RHEL 8 (Milestone 2)
We would like to automate our release pipeline to speed up our release process for agents that are available to us RHEL 8

### Milestone 2
- [ ] Use two agents RHEL 8 and deploy two containers 1. to pack the CLI and run the integration tests 2. Conjur server

### PR DESCRIPTION
**Pull request by eranha**
### What does this PR do?
Run CLI tests against appliance running on RHEL
- _What's changed? Why were these changes made?_
  - Added a new step in Jenkins file 
  - Added a new docker file for test on RHEL
  - Docker compose file
    - Updated Conjur image version
    - Added Appliance container
    - Added cli container
- _How should the reviewer approach this PR, especially if manual tests are required?_
  - No regression

### What ticket does this PR close?
Resolves #181 

#### Change log
- [ ] The CHANGELOG has been updated, or
- [x] This PR does not include user-facing changes and doesn't require a CHANGELOG update

#### Test coverage
- [ ] This PR includes new unit and integration tests to go with the code changes, or
- [ ] The changes in this PR do not require tests

#### Documentation
- [ ] Docs (e.g. `README`s) were updated in this PR, and/or there is a follow-on issue to update docs, or
- [x] This PR does not require updating any documentation

### CODE REVIEW

File: bin/test_rhel8_integration [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+echo "Generating certificate..."
+docker-compose up openssl
+
+echo "Starting Conjur..."
+docker-compose up -d conjur-appliance
+
+
+echo "Configuring Conjur..."
+export ADMIN_PASSWORD=P@S5-ITAbXJlxH1MU
```
Comment (mbenita-Cyberark) [inline bin/test_rhel8_integration:30]: I think it will be better to generate a key here dynamically so the password will not be exposed to the outside world

File: bin/test_rhel8_integration [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+export PYTHON_KEYRING_BACKEND=keyring.backends.SecretService.Keyring
```
Comment (mbenita-Cyberark) [inline bin/test_rhel8_integration:31]: not sure this is needed anymore

**Review by mbenita-Cyberark — COMMENTED** [review]
just went over to have better understanding on the task and left some comments

File: docker-compose.yml [hunk 3]
```diff
@@ -29,7 +29,7 @@ services:
       POSTGRES_HOST_AUTH_METHOD: trust
 
   conjur:
-    image: cyberark/conjur:1.9
+    image: cyberark/conjur:1.11.4
```
Comment (sigalsax) [inline docker-compose.yml:32]: this should be a separate PR IMO

File: test/configure_dbus.sh [hunk 4]
```diff
@@ -1,20 +1,34 @@
 #!/bin/bash
 
 echo "Configuring D-bus environment"
-echo 'eval "$(dbus-launch --sh-syntax)"' > /dbus.sh
+
+echo "#!/bin/bash -ex" >> /dbus.sh
```
Comment (mbenita-Cyberark) [inline test/configure_dbus.sh:5]: @sigalsax I think this will be a good idea to change `/dbus.sh` file name into `tests_executer`
and the `test/configure_dbus.sh ` filename is also not relevant I think.
WDYT?

**Review by mbenita-Cyberark — COMMENTED** [review]
Amazing this is now works

File: CentOS-Linux-AppStream.repo [hunk 5]
```diff
@@ -0,0 +1,20 @@
+# CentOS-Linux-AppStream.repo
+#
+# The mirrorlist system uses the connecting IP address of the client and the
+# update status of each mirror to pick current mirrors that are geographically
+# close to the client.  You should use this for CentOS updates unless you are
+# manually picking other mirrors.
+#
+# If the mirrorlist does not work for you, you can try the commented out
+# baseurl line instead.
```
Comment (sigalsax) [inline CentOS-Linux-AppStream.repo:9]: do we need this comment? because I don't understand its relation to being able to add our dependencies

File: Dockerfile.test.rhel8 [hunk 6]
```diff
@@ -0,0 +1,38 @@
+FROM registry.access.redhat.com/ubi8/ubi:8.3
+
+ENV INSTALL_DIR=/opt/conjur-api-python3
+
+# Copy below repo file to enable installation of gnome-keyring and dbus-x11
+COPY ./CentOS-Linux-AppStream.repo /etc/yum.repos.d/CentOS-Linux-AppStream.repo
+
+RUN yum --disableplugin=subscription-manager -y install -y \
+            bash \
```
Comment (sigalsax) [inline Dockerfile.test.rhel8:9]: indentation. I think all the following dependencies should be right below the `\` in the previous line or `install`

File: Jenkinsfile [hunk 7]
```diff
@@ -49,6 +49,18 @@ pipeline {
         }
       }
     }
+
+    stage('RHEL8 Integration tests') {
+      steps {
+        sh './bin/test_rhel8_integration'
```
Comment (sigalsax) [inline Jenkinsfile:55]: in terms of naming it looks cleaner imo to have the platform at the end so it becomes `test_rhel8_integration` -> `test_integration_rhel8`. That way, at a first glance it is clear this script is for rhel8

File: Jenkinsfile [hunk 8]
```diff
@@ -98,4 +110,4 @@ pipeline {
       cleanupAndNotify(currentBuild.currentResult)
     }
   }
-}
+}
```
Comment (sigalsax) [inline Jenkinsfile:113]: new line :)

File: bin/build_integrations_tests_runner [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+  source venv/bin/activate
+
+  echo "Installing requirements for tests..."
+  pip3 install -r requirements.txt
+
+  pyinstaller -F /opt/conjur-api-python3/test/util/test_runners/integrations_tests_runner.py
+  deactivate
+}
+
+_build_test_integration_exec
```
Comment (sigalsax) [inline bin/build_integrations_tests_runner:15]: new line :)

File: bin/build_integrations_tests_runner [hunk 10]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash -e
+
+function _build_test_integration_exec() {
+  echo "Dropping into virtual environment"
```
Comment (sigalsax) [inline bin/build_integrations_tests_runner:4]:
```suggestion
  echo "Dropping into virtual environment..."
```

Add the `...` because it is an action that is happening

File: bin/test_rhel8_integration (hunk 1 above)
Comment (sigalsax) [reply bin/test_rhel8_integration:30]: ^ agree

File: bin/test_rhel8_integration (hunk 2 above)
Comment (sigalsax) [reply bin/test_rhel8_integration:31]: agreed ^

File: bin/test_rhel8_integration [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+export DEBUG=$DEBUG
+export CONJUR_AUTHN_API_KEY=P@S5-ITAbXJlxH1MU
```
Comment (sigalsax) [inline bin/test_rhel8_integration:33]: I dont think this should be hardcoded as @mbenita-Cyberark  mentioned

File: bin/test_rhel8_integration [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+export SERVER_MODE="appliance"
+export ACCOUNT="dev"
+export LOGIN="admin"
+export TEST_HOSTNAME=conjur-appliance
+
+# Docker-compose will pick this file up when starting containers
+# so we will have these variables in the container
+cat <<ENV > .env
+CONJUR_AUTHN_API_KEY=$ADMIN_PASSWORD
+PYTHON_KEYRING_BACKEND=$PYTHON_KEYRING_BACKEND
```
Comment (sigalsax) [inline bin/test_rhel8_integration:43]: remove

File: bin/test_rhel8_integration [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+
+echo "Configuring Conjur..."
+export ADMIN_PASSWORD=P@S5-ITAbXJlxH1MU
+export PYTHON_KEYRING_BACKEND=keyring.backends.SecretService.Keyring
+export DEBUG=$DEBUG
+export CONJUR_AUTHN_API_KEY=P@S5-ITAbXJlxH1MU
+export SERVER_MODE="appliance"
+export ACCOUNT="dev"
+export LOGIN="admin"
+export TEST_HOSTNAME=conjur-appliance
```
Comment (sigalsax) [inline bin/test_rhel8_integration:37]: I worry about this because we have this in constants.py. I agree that this value should be here instead of in the constants.py so in the test_integration file can you add `export TEST_HOSTNAME=conjur-https`

File: docker-compose.yml [hunk 14]
```diff
@@ -78,3 +78,36 @@ services:
...
+
+  test-rhel8:
+    privileged: true
+    build:
+      context: .
+      dockerfile: Dockerfile.test.rhel8
+    command: ['sleep', '999d']
+    environment:
+      CONJUR_AUTHN_API_KEY: Null
+      PYTHON_KEYRING_BACKEND: Null
```
Comment (sigalsax) [inline docker-compose.yml:102]: remove

File: test/configure_dbus.sh (hunk 4 above)
Comment (sigalsax) [reply test/configure_dbus.sh:5]: > I think this will be a good idea to change /dbus.sh file name into tests_executer

why tests_executor? they all env vars saved are dbus related 

>  the test/configure_dbus.sh filename is also not relevant I think.

If anything this should be test_executor IMO because we are setting up the environment and executing the tests @mbenita-Cyberark

File: test/test_integration_list.py [hunk 15]
```diff
@@ -159,20 +159,20 @@ def test_list_limit_long_returns_same_number_of_resources(self):
     @integration_test(True)
     def test_list_limit_invalid_param_returns_empty_list(self):
         output = self.invoke_cli(self.cli_auth_params, ['list', '-l', '0'], exit_code=1)
-        self.assertIn("500 Server Error", output)
+        self.assertIn("422 Client Error", output)
```
Comment (sigalsax) [inline test/test_integration_list.py:162]: why was this changed?

File: Dockerfile.test.rhel8 [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+            git \
+            jq \
+            libffi-devel \
+            python3 \
+            python3-devel \
+            python3-pip \
+            gnome-keyring \
+            dbus-x11 \
+            procps \
+         && yum --disableplugin=subscription-manager clean all
```
Comment (sigalsax) [inline Dockerfile.test.rhel8:23]: what does this do?

**Comment by sigalsax** [comment]
@eranha  left some comments :) fantastic work!

File: docker-compose.yml (hunk 3 above)
Comment (eranha) [reply docker-compose.yml:32]: we changes the tests so it woll work with appliance it will be too much hustle now

File: test/test_integration_list.py (hunk 15 above)
Comment (eranha) [reply test/test_integration_list.py:162]: Because of the change in conjur version in appliance

File: Dockerfile.test.rhel8 (hunk 16 above)
Comment (eranha) [reply Dockerfile.test.rhel8:23]: prevents the repo mngr to go out

File: docker-compose.yml (hunk 3 above)
Comment (sigalsax) [reply docker-compose.yml:32]: dont understand how it is related. This is related to ubuntu integration tests, not rhel tests

File: Dockerfile.test.rhel8 [hunk 17]
```diff
@@ -0,0 +1,38 @@
+FROM registry.access.redhat.com/ubi8/ubi:8.3
+
+ENV INSTALL_DIR=/opt/conjur-api-python3
+
+# Copy below repo file to enable installation of gnome-keyring and dbus-x11
+COPY ./CentOS-Linux-AppStream.repo /etc/yum.repos.d/CentOS-Linux-AppStream.repo
+
+RUN yum --disableplugin=subscription-manager -y \
+    install -y  bash \
```
Comment (sigalsax) [inline Dockerfile.test.rhel8:9]: indentation wasn't fixed 🙃

File: Dockerfile.test [hunk 18]
```diff
@@ -23,6 +23,6 @@ RUN pip3 install -r requirements.txt
 
 COPY . $INSTALL_DIR
 
-COPY ./test/configure_dbus.sh /configure_dbus.sh
+COPY ./test/generate_tests_exec_script.sh /generate_tests_exec_script.sh
 
-ENTRYPOINT ["./test/configure_dbus.sh"]
+ENTRYPOINT ["./test/generate_tests_exec_script.sh"]
```
Comment (sigalsax) [inline Dockerfile.test:28]: I dont think we need to say `_script` at the end of the naming because it is clear it is a script.

Comment (sigalsax) [reply Dockerfile.test:28]: what about just `test_executor.sh`

File: bin/test_integration [hunk 19]
```diff
@@ -48,12 +49,12 @@ rm -rf $CURRENT_DIR/output/*
 # Hint: '--privileged' is added here otherwise we will get operation
 # not permitted when attempting to do anything related to dbus
 if [[ "$DEBUG" == "true" ]]; then
-  docker-compose exec --privileged test /bin/sh -c " /dbus.sh" $@
+  docker-compose exec --privileged test /bin/sh -c " /tests_executor.sh" $@
```
Comment (sigalsax) [inline bin/test_integration:52]: I don't think test_executor is a good name here because the script is strictly for dbus configurations . I would call it `configure_dbus.sh`

@mbenita-Cyberark @eranha

File: bin/test_integration_rhel8 [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+  { </dev/urandom LC_ALL=C grep -ao '[A-Za-z0-9]' \
+        | head -n$((RANDOM % 12 + 9))
+    echo ${chars:$((RANDOM % ${#chars})):1}   # Random special char.
+  } \
+    | shuf \
+    | tr -d '\n'
+}
+
+echo "Configuring Conjur..."
+export ADMIN_PASSWORD=$(_random_password)
```
Comment (sigalsax) [inline bin/test_integration_rhel8:39]: please pay mind to forbidden character combinations.

File: docker-compose.yml (hunk 3 above)
Comment (sigalsax) [reply docker-compose.yml:32]: @eranha ?

File: test/generate_tests_exec_script.sh [hunk 21]
```diff
@@ -0,0 +1,61 @@
+#!/bin/bash
+
+###################################
+# Docker image entry point script #
+###################################
+
+script_file=/tests_executor.sh
```
Comment (sigalsax) [inline test/generate_tests_exec_script.sh:7]: again, would call it configure_dbus.sh

File: test/generate_tests_exec_script.sh [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+
+mkdir -p ~/.cache
+mkdir -p ~/.local/share/keyrings
+
+# Write the tests runner script that configures DBUS and runs
+# the tests.
+cat > $script_file <<'endmsg'
```
Comment (sigalsax) [inline test/generate_tests_exec_script.sh:14]: nice!

File: test/generate_tests_exec_script.sh [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+function append_to_file() {
+  echo "$1" >> $script_file
+}
+
+if [ "$DEBUG" == "true" ]; then
+  echo "bash" >> $script_file
+elif [ "$SERVER_MODE" == "appliance" ]; then
+    # Server mode is Conjur Appliance
+    # Add tests runner executable command
+    echo "SERVER_MODE is appliance, building the test integration executable..."
```
Comment (sigalsax) [inline test/generate_tests_exec_script.sh:49]: Maybe `Running testing against $SERVER_MODE. Building the test integration executable..

Comment (sigalsax) [reply test/generate_tests_exec_script.sh:49]: I also wouldn't hardcode appliance for better dbugging

File: test/generate_tests_exec_script.sh [hunk 24]
```diff
@@ -0,0 +1,61 @@
...
+    source /build_integrations_tests_runner
+    append_to_file "$(_tests_runner_cmd)"
```
Comment (sigalsax) [inline test/generate_tests_exec_script.sh:51]: Can you explain a bit how the script is triggered? seems like we are only appending to the test_exector.sh script but missing how the script is being executed

File: test/generate_tests_exec_script.sh [hunk 25]
```diff
@@ -0,0 +1,61 @@
...
+  echo "$1" >> $script_file
+}
+
+if [ "$DEBUG" == "true" ]; then
+  echo "bash" >> $script_file
+elif [ "$SERVER_MODE" == "appliance" ]; then
+    # Server mode is Conjur Appliance
+    # Add tests runner executable command
+    echo "SERVER_MODE is appliance, building the test integration executable..."
+    source /build_integrations_tests_runner
```
Comment (sigalsax) [inline test/generate_tests_exec_script.sh:50]: why do we source here?

File: bin/test_integration_rhel8 [hunk 26]
```diff
@@ -0,0 +1,90 @@
...
+export LOGIN="admin"
+export TEST_HOSTNAME=conjur-appliance
+
+# Docker-compose will pick this file up when starting containers
+# so we will have these variables in the container
+cat <<ENV > .env
+CONJUR_AUTHN_API_KEY=$ADMIN_PASSWORD
+DEBUG=$DEBUG
+SERVER_MODE=$SERVER_MODE
+TEST_HOSTNAME=$TEST_HOSTNAME
```
Comment (sigalsax) [inline bin/test_integration_rhel8:53]: my comment disappeared but I suggested we also have the same logic in the test_integration for the regular test_integration file in the ubuntu flow where we bring over TEST_HOSTNAME

File: docker-compose.yml (hunk 3 above)
Comment (eranha) [reply docker-compose.yml:32]: in conjur 1.9 the error was server error 500 and it changed in appliance because of different conjur version

File: Dockerfile.test (hunk 18 above)
Comment (eranha) [reply Dockerfile.test:28]: `test_executor.sh` is the name of the file being generated.

File: bin/test_integration (hunk 19 above)
Comment (eranha) [reply bin/test_integration:52]: The script runs the tests at the end.

File: docker-compose.yml (hunk 3 above)
Comment (sigalsax) [reply docker-compose.yml:32]: so I still think this and the tests you rightfully corrected should be in a separate PR. This is not a minor task in my opinion that should be coupled with this PR because it effects our test plan

File: Dockerfile.test (hunk 18 above)
Comment (sigalsax) [reply Dockerfile.test:28]: I know but the file being generated should be configure_dbus because that is what that generated file is essentially doing

Comment (sigalsax) [reply Dockerfile.test:28]: it has all the logic specific to dbus and nothing about the actual tests so my suggestion here is that the actual script that triggers the tests should be `test_executor.sh` and the dbus generated file should be `configure_dbus.sh`

**Review by sigalsax — APPROVED** [review]
LGTM! Please squash/merge and replace your commits with a single more descriptive message
